graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -382,6 +387,12 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
         in_topmost_optional_block = False
 
         edge_traversal_is_optional = optional_directive is not None
+
+        # This is true for any vertex expanded within an @optional scope.
+        within_optional_scope = 'optional' in context and not edge_traversal_is_optional
+        # Currently @optional is not allowed within @optional.
```

Comment:
It's nice to keep comments right above the thing they are referring to. This is because right now, someone could accidentally put something on line 395 and make it confusing what the comment refers to.

I'd move lines 393-394 to right above 392.